File: src/voice_tools/tools/recording_qa/review.py
"""只接受显式人工标签；与合成夹具的期望结果分开管理。"""
import csv
from datetime import datetime, timezone
import json
import math
from pathlib import Path
import re

from voice_tools.core.files import read_json, sha256, write_json
from .reports import LABELS, REVIEW_FIELDS, EXTRA_REVIEW_FIELDS

DECISIONS = {"missing", "delayed", "audible", "exclude", "uncertain"}


def result_index(path):
    index = {}
    with Path(path).open(encoding="utf-8") as stream:
        for line in stream:
            if not line.strip():
                continue
            record = json.loads(line)
            if not isinstance(record, dict) or record.get("schema_version") != "1.0":
                raise ValueError("不支持的结果 schema_version")
            if "error" in record:
                continue
            if (not {"sample_id", "audio_sha256", "result"} <= record.keys()
                    or not isinstance(record["result"], dict)
                    or not isinstance(record["result"].get("opportunities"), list)):
                raise ValueError("结果记录缺少样本标识或应答机会列表")
            for field in ("sample_id", "audio_sha256"):
                if not isinstance(record[field], str) or not re.fullmatch(r"[0-9a-f]{64}", record[field]):
                    raise ValueError("结果中的样本标识或摘要无效")
            for item in record["result"]["opportunities"]:
                if not isinstance(item, dict) or not {"id", "at_s", "observed_until_s", "status"} <= item.keys():
                    raise ValueError("结果中的应答机会缺少必要字段")
                if not isinstance(item["id"], str) or not item["id"] or not isinstance(item["status"], str) or item["status"] not in LABELS:
                    raise ValueError("结果中的机会标识或状态无效")
                for field in ("at_s", "observed_until_s"):
                    value = item[field]
                    if type(value) not in (int, float) or not math.isfinite(value) or value < 0:
                        raise ValueError("结果中的机会时间无效")
                if item["observed_until_s"] < item["at_s"]:
                    raise ValueError("结果中的机会时间倒置")
                key = (record["sample_id"], item["id"])
                value = {**item, "audio_sha256": record["audio_sha256"],
                         "timeout_s": record["result"].get("config", {}).get("timeout_s")}
                if key in index and index[key] != value:
                    raise ValueError("同一样本/应答机会出现相互冲突的结果")
                index[key] = value
    return index


def new_file(path):
    path = Path(path)
    if path.exists():
        raise ValueError(f"拒绝覆盖已有文件：{path}")
    path.parent.mkdir(parents=True, exist_ok=True)
    return path


def timestamp(text):
    if not isinstance(text, str):
        raise ValueError("reviewed_at 必须是时间字符串")
    try:
        value = datetime.fromisoformat(text.replace("Z", "+00:00"))
    except ValueError as error:
        raise ValueError("reviewed_at 必须为含时区的 ISO 8601 时间") from error
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("reviewed_at 必须包含时区")
    return value.isoformat()


def check_window(label, predicted):
    if label["audio_sha256"] != predicted["audio_sha256"]:
        raise ValueError("标签的录音摘要与结果不一致")
    for key in ("at_s", "observed_until_s"):
        if isinstance(label[key], bool):
            raise ValueError("标签窗口不能是布尔值")
        try:
            value = float(label[key])
        except (TypeError, ValueError) as error:
            raise ValueError(f"标签 {key} 必须是有效时间") from error
        if not math.isfinite(value) or abs(value - predicted[key]) > 1e-6:
            raise ValueError("标签窗口与结果不一致；请使用固定事件文件对齐后再评估")


def extra_fields(row):
    extra = {}
    for field in EXTRA_REVIEW_FIELDS:
        value = row.get(field)
        if value is None or value == "":
            continue
        if field in ("first_audible_s", "deadline_s"):
            try:
                number = float(value)
            except (ValueError, TypeError) as error:
                raise ValueError(f"{field} 须为有限数值") from error
            if isinstance(value, bool) or not math.isfinite(number) or number < 0 or (field == "deadline_s" and number <= 0):
                raise ValueError(f"{field} 须为有效正时间")
            extra[field] = number
        elif field == "expected_response":
            if value not in (True, False, "true", "false") or type(value) not in (str, bool):
                raise ValueError("expected_response 须为 true/false")
            extra[field] = value is True or value == "true"
        else:
            if not isinstance(value, str) or not value.strip():
                raise ValueError(f"{field} 须为非空字符串")
            extra[field] = value.strip()
    if "first_audible_s" in extra:
        if not float(row["at_s"]) <= extra["first_audible_s"] < float(row["observed_until_s"]):
            raise ValueError("首次可听回答须在标注窗口内")
        if row["decision"] == "missing":
            raise ValueError("缺失回答标签不能同时记录可听回答时间")
    if "deadline_s" in extra and not extra.get("policy_id"):
        raise ValueError("应答时限须同时记录 policy_id")
    if "first_audible_s" in extra and "deadline_s" in extra and row["decision"] in ("audible", "delayed"):
        delayed = extra["first_audible_s"] - float(row["at_s"]) + 1e-8 >= extra["deadline_s"]
        if (row["decision"] == "delayed") != delayed:
            raise ValueError("回答时间与人工迟答标签/时限不一致")
    if extra.get("expected_response") is False and row["decision"] not in ("exclude", "uncertain"):
        raise ValueError("无需应答不能标为缺失、迟答或有效回答")
    if extra.get("split") not in (None, "calibration", "validation") or (extra.get("split") and not extra.get("group_id")):
        raise ValueError("样本集合须为 calibration/validation 并提供 group_id")
    return extra


def validate_splits(labels):
    assignments = {}
    for label in labels:
        if not label.get("split"):
            continue
        for key in (("group", label["group_id"]), ("audio", label["audio_sha256"])):
            if key in assignments and assignments[key] != label["split"]:
                raise ValueError("同一来源分组或录音不能跨校准集和留出验证集")
            assignments[key] = label["split"]


def promote(review_path, results_path, output, dataset_kind):
    if dataset_kind not in ("synthetic", "real"):
        raise ValueError("dataset_kind 只能为 synthetic 或 real")
    index = result_index(results_path)
    labels, seen = [], set()
    with Path(review_path).open(encoding="utf-8-sig", newline="") as stream:
        reader = csv.DictReader(stream)
        if not set(REVIEW_FIELDS) <= set(reader.fieldnames or []):
            raise ValueError("人工复核 CSV 缺少必需列，请从分析报告导出")
        for row in reader:
            # CSV 输出为电子表格防注入加的前缀，只在格式相符时还原。
            row = {k: (v[1:] if v.startswith("'") and v[1:].lstrip().startswith(("=", "+", "-", "@")) else v)
                   for k, v in row.items() if k is not None and v is not None}
            decision = row.get("decision", "").strip()
            if not decision:
                if any(row.get(k, "").strip() for k in ("reviewer", "reviewed_at", "notes", *EXTRA_REVIEW_FIELDS)):
                    raise ValueError("存在已填写复核信息但缺少 decision 的半成品标签")
                continue
            if decision not in DECISIONS or not row.get("reviewer", "").strip():
                raise ValueError("人工标签须有合法 decision 和非空 reviewer")
            reviewed_at = timestamp(row.get("reviewed_at", ""))
            key = (row.get("sample_id"), row.get("opportunity_id"))
            if key not in index or key in seen:
                raise ValueError("标签引用了不存在或重复的样本/应答机会")
            check_window(row, index[key])
            seen.add(key)
            labels.append({"sample_id": key[0], "opportunity_id": key[1],
                           "audio_sha256": row["audio_sha256"],
                           "at_s": float(row["at_s"]), "observed_until_s": float(row["observed_until_s"]),
                           "decision": decision, "reviewer": row["reviewer"].strip(),
                           "reviewed_at": reviewed_at, "notes": row.get("notes", ""), **extra_fields(row)})
    if not labels:
        raise ValueError("没有人工复核标签；不能把空白或自动结果晋升为黄金集")
    validate_splits(labels)
    golden = {"schema_version": "1.1", "label_source": "human_review", "dataset_kind": dataset_kind,
              "source_results_sha256": sha256(results_path),
              "created_at": datetime.now(timezone.utc).isoformat(), "labels": labels}
    write_json(new_file(output), golden)
    return golden


def evaluate(golden_path, results_path):
    golden, index = read_json(golden_path), result_index(results_path)
    if not isinstance(golden, dict) or golden.get("schema_version") not in ("1.0", "1.1") or golden.get("label_source") != "human_review":
        raise ValueError("仅接受人工复核黄金集 schema 1.0/1.1")
    if golden.get("dataset_kind") not in ("synthetic", "real"):
        raise ValueError("黄金集必须声明 synthetic / real")
    if not isinstance(golden.get("labels"), list) or not golden["labels"]:
        raise ValueError("黄金集标签为空或无效")
    seen = set()
    for label in golden["labels"]:
        if not isinstance(label, dict) or not {"sample_id", "opportunity_id", "audio_sha256", "at_s", "observed_until_s", "decision", "reviewer", "reviewed_at"} <= label.keys():
            raise ValueError("黄金标签缺少必要字段")
        if not all(isinstance(label[field], str) for field in ("sample_id", "opportunity_id", "decision")):
            raise ValueError("黄金标签标识或判断类型无效")
        key = (label["sample_id"], label["opportunity_id"])
        if key in seen or key not in index:
            raise ValueError("黄金标签重复或缺少对应预测；不能静默跳过")
        seen.add(key)
        decision = label["decision"]
        if decision not in DECISIONS or not isinstance(label["reviewer"], str) or not label["reviewer"].strip():
            raise ValueError("黄金标签不是有效的人工复核记录")
        timestamp(label["reviewed_at"])
        check_window(label, index[key])
        extra = extra_fields(label)
        if "deadline_s" in extra and index[key].get("timeout_s") != extra["deadline_s"]:
            raise ValueError("黄金标签与检测结果的应答时限不一致")
    validate_splits(golden["labels"])
    from .metrics import detailed_metrics
    metrics = detailed_metrics(golden["labels"], index)
    if not metrics["evaluated"] and not metrics["abstained"]:
        raise ValueError("没有可计分的明确标签，无法计算指标")
    return {"schema_version": "1.1", "dataset_kind": golden["dataset_kind"], **metrics,
            "notice": "仅衡量当前人工标签集；synthetic 不代表真实准确率。观察不足/证据不足/算法排除单列为无法判定，不算通过。Wilson 区间按机会计算，相关通话应按组独立验证。"}

File: src/voice_tools/tools/recording_qa/review.js
'use strict';
(() => {
  const data = JSON.parse(document.getElementById('reviewData').textContent);
  const $ = (id) => document.getElementById(id);
  const labels = new Map(),
    entries = [],
    byKey = new Map();
  const extra = [
    'first_audible_s',
    'deadline_s',
    'policy_id',
    'expected_response',
    'scenario',
    'line_id',
    'group_id',
    'split'
  ];
  let current = null,
    filtered = [],
    dirty = false,
    formDirty = false;
  const key = (row) => JSON.stringify([row.sample_id, row.opportunity_id]);
  const player = $('player');
  const fileFilter = createReviewFileFilter(data.records, (record) => record.sample_id, refreshFilters);
  function message(text) {
    $('message').textContent = text;
  }
  const waveform = createReviewWaveform(
    player,
    (start, end) => {
      playback.setRange(start, end);
    },
    message
  );
  const playback = createReviewPlayback(player, {
    duration: () => current?.record.result.duration_s,
    seek: (value) => waveform.setTime(value),
    rangeChanged: (start, end) => waveform.syncRange(start, end),
    error: message
  });
  for (const record of data.records) {
    if (!record.result) continue;
    for (const op of record.result.opportunities) {
      const row = {
        sample_id: record.sample_id,
        audio_sha256: record.audio_sha256,
        opportunity_id: op.id,
        at_s: op.at_s,
        observed_until_s: op.observed_until_s,
        status: op.status
      };
      const entry = { record, op, row, key: key(row) };
      entries.push(entry);
      byKey.set(entry.key, entry);
    }
  }
  for (const status of [...new Set(entries.map((e) => e.op.status))])
    $('statusFilter').add(new Option(data.labels[status], status));
  function progress() {
    $('progress').textContent =
      `${data.summary.files} 个录音 · ${entries.length} 个机会 · ${labels.size} 条已填写 · ${filtered.length} 条符合筛选` +
      (dirty || formDirty ? ' · 有未导出的改动' : ' · 无未导出的改动');
  }
  function refresh() {
    const queue = $('queue'),
      scrollTop = queue.scrollTop;
    const focusedKey = queue.contains(document.activeElement)
      ? document.activeElement.dataset.key
      : null;
    filtered = entries.filter(
      (e) =>
        fileFilter.matches(e.record) &&
        (!$('statusFilter').value || e.op.status === $('statusFilter').value) &&
        (!$('evidenceFilter').value || e.op.evidence_level === $('evidenceFilter').value) &&
        (!$('unreviewed').checked || !labels.has(e.key))
    );
    queue.replaceChildren();
    for (const entry of filtered) {
      const button = document.createElement('button');
      button.type = 'button';
      button.setAttribute('aria-current', String(current === entry));
      button.dataset.key = entry.key;
      button.textContent = `${labels.has(entry.key) ? '✓ ' : ''}${entry.record.input.split(/[\\/]/).pop()} · ${entry.op.id}`;
      const small = document.createElement('small');
      small.textContent = `${entry.op.at_s.toFixed(2)} s · ${data.labels[entry.op.status]}`;
      button.append(small);
      button.onclick = () => select(entry);
      $('queue').append(button);
    }
    if (!filtered.length) {
      const p = document.createElement('p');
      p.textContent = '没有符合条件的机会。';
      $('queue').append(p);
      if (!formDirty) {
        playback.pause();
        current = null;
        $('detail').hidden = true;
      }
    }
    queue.scrollTop = scrollTop;
    if (focusedKey)
      [...queue.children]
        .find((button) => button.dataset.key === focusedKey)
        ?.focus({ preventScroll: true });
    progress();
  }
  function updateQueueSelection() {
    const queue = $('queue'),
      hadFocus = queue.contains(document.activeElement);
    for (const button of queue.querySelectorAll('button')) {
      const selected = button.dataset.key === current.key;
      button.setAttribute('aria-current', String(selected));
      if (!selected) continue;
      // Scroll only the list so navigation does not move the review form.
      const item = button.getBoundingClientRect(),
        viewport = queue.getBoundingClientRect();
      const top = viewport.top + queue.clientTop + 3,
        bottom = viewport.top + queue.clientTop + queue.clientHeight - 3;
      if (item.top < top || item.height > bottom - top) queue.scrollTop += item.top - top;
      else if (item.bottom > bottom) queue.scrollTop += item.bottom - bottom;
      if (hadFocus) button.focus({ preventScroll: true });
    }
  }
  function saveDraft() {
    if (!formDirty) return true;
    message('当前表单尚未记下，请先记下标注或点击放弃表单修改。');
    return false;
  }
  function select(entry) {
    if (!saveDraft()) return;
    playback.pause();
    current = entry;
    $('detail').hidden = false;
    const { record, op } = entry,
      result = record.result;
    $('name').textContent = `${record.input.split(/[\\/]/).pop()} · ${op.id}`;
    $('metadata').textContent =
      `${result.sample_rate} Hz · ${result.duration_s.toFixed(2)} 秒 · 标注窗口 ${op.at_s.toFixed(2)}–${op.observed_until_s.toFixed(2)} 秒`;
    $('status').textContent = `${data.labels[op.status]} · ${data.evidence[op.evidence_level]}`;
    $('warning').textContent = result.warnings.join('；') || '活动不代表有效回答，请试听后判断。';
    for (const [id, channel] of [
      ['leftTitle', 0],
      ['rightTitle', 1]
    ]) {
      $(id).querySelector('.role-name').textContent =
        result.config.system_channel === channel ? 'AI' : '用户';
      $(id).querySelector('.role-meta').textContent =
        `${channel === 0 ? '左' : '右'}声道\n${result.channel_verified ? '已核实' : '未核实'}`;
    }
    $('start').value = Number(Math.max(0, op.at_s - 2).toFixed(6));
    $('end').value = Number(Math.min(result.duration_s, op.observed_until_s + 1).toFixed(6));
    for (const id of ['start', 'end']) $(id).removeAttribute('aria-invalid');
    $('seek').max = result.duration_s;
    $('time').textContent =
      `${Number($('start').value).toFixed(2)} / ${result.duration_s.toFixed(2)} 秒`;
    const label = labels.get(entry.key) || {};
    $('decision').value = label.decision || '';
    $('notes').value = label.notes || '';
    if (label.reviewer) $('reviewer').value = label.reviewer;
    for (const id of extra) $(id).value = label[id] ?? '';
    const clip = record.clips && record.clips[op.id];
    $('clip').hidden = $('clipMeta').hidden = !clip;
    if (clip) {
      $('clip').href = clip.audio;
      $('clipMeta').href = clip.metadata;
    }
    switchSource(false);
    waveform.load(entry);
    updateQueueSelection();
    progress();
  }
  function switchSource(keepTime = true) {
    if (!current) return;
    const sources = current.record.playback_sources || {};
    // A previous recording may not offer the same selected channel.
    if (!sources[$('channel').value])
      $('channel').value = sources.both ? 'both' : sources.left ? 'left' : 'both';
    const source = sources[$('channel').value];
    for (const option of $('channel').options) option.disabled = !sources[option.value];
    player.hidden = true;
    waveform.syncChannel($('channel').value);
    playback.setSource(source, keepTime);
    message(source ? '' : '未附带音频，无法试听。波形和标注仍可使用。');
  }
  player.addEventListener('error', () => {
    if (player.getAttribute('src'))
      message('音频无法播放，请确认报告 audio 目录完整或使用本地 HTTP 打开。');
  });
  $('channel').onchange = () => switchSource();
  function validate(row) {
    const entry = byKey.get(key(row));
    if (!entry) throw Error('标签不属于本批次录音/事件/应答机会');
    if (row.audio_sha256 !== entry.row.audio_sha256) throw Error('音频摘要不匹配');
    for (const id of ['at_s', 'observed_until_s'])
      if (
        row[id] === '' ||
        !Number.isFinite(Number(row[id])) ||
        Math.abs(Number(row[id]) - entry.row[id]) > 1e-6
      )
        throw Error('标注窗口不匹配');
    if (
      !['missing', 'delayed', 'audible', 'exclude', 'uncertain'].includes(row.decision) ||
      !row.reviewer?.trim()
    )
      throw Error('需要明确判断和复核人');
    if (
      !/(Z|[+-]\d\d:\d\d)$/.test(row.reviewed_at || '') ||
      Number.isNaN(Date.parse(row.reviewed_at))
    )
      throw Error('复核时间须有效且含时区');
    if (
      row.first_audible_s !== '' &&
      row.first_audible_s != null &&
      (!Number.isFinite(Number(row.first_audible_s)) ||
        Number(row.first_audible_s) < entry.op.at_s ||
        Number(row.first_audible_s) >= entry.op.observed_until_s)
    )
      throw Error('首次可听回答须在标注窗口内');
    if (row.first_audible_s !== '' && row.first_audible_s != null && row.decision === 'missing')
      throw Error('缺失回答不能同时记录可听回答时间');
    if (
      row.deadline_s &&
      (!Number.isFinite(Number(row.deadline_s)) ||
        Number(row.deadline_s) <= 0 ||
        !row.policy_id?.trim())
    )
      throw Error('应答时限须为正数并填写策略版本');
    if (
      row.deadline_s &&
      row.first_audible_s !== '' &&
      row.first_audible_s != null &&
      ['audible', 'delayed'].includes(row.decision) &&
      Number(row.first_audible_s) - entry.op.at_s + 1e-8 >= Number(row.deadline_s) !==
        (row.decision === 'delayed')
    )
      throw Error('回答时间与人工迟答标签/时限不一致');
    if (row.split && (!['calibration', 'validation'].includes(row.split) || !row.group_id?.trim()))
      throw Error('样本集合需合法且填写来源分组 ID');
    if (row.expected_response && !['true', 'false'].includes(row.expected_response))
      throw Error('是否应答须为 true/false');
    if (row.expected_response === 'false' && !['exclude', 'uncertain'].includes(row.decision))
      throw Error('无需应答应标为无需应答或无法判断');
    return entry;
  }
  $('reviewForm').addEventListener('input', () => {
    formDirty = true;
    progress();
  });
  $('reviewForm').onsubmit = (event) => {
    event.preventDefault();
    if (!current) return;
    const row = {
      ...current.row,
      decision: $('decision').value,
      reviewer: $('reviewer').value.trim(),
      reviewed_at: new Date().toISOString(),
      notes: $('notes').value
    };
    for (const id of extra) row[id] = $(id).value.trim();
    try {
      validate(row);
      labels.set(current.key, row);
      dirty = true;
      formDirty = false;
      refresh();
      reconcileSelection();
      message('标注已记在本页，请导出 CSV 后再离开。');
    } catch (e) {
      message(e.message);
    }
  };
  function move(delta) {
    const index = filtered.indexOf(current);
    if (filtered.length)
      select(filtered[Math.max(0, Math.min(filtered.length - 1, index + delta))]);
  }
  $('previous').onclick = () => move(-1);
  $('next').onclick = () => move(1);
  function reconcileSelection() {
    if (formDirty) return;
    if (filtered.length && !filtered.includes(current)) select(filtered[0]);
    else if (!filtered.length) {
      playback.pause();
      current = null;
      $('detail').hidden = true;
    }
  }
  $('discard').onclick = () => {
    formDirty = false;
    refresh();
    reconcileSelection();
    if (current) select(current);
    message('已放弃未记下的修改，并恢复当前筛选结果。');
  };
  function refreshFilters() {
    refresh();
    if (formDirty && !filtered.includes(current))
      message('筛选已更新；当前表单有未记下的修改，记下或放弃后再切换录音。');
    else reconcileSelection();
  }
  for (const id of ['directoryFilter', 'fileFilter', 'statusFilter', 'evidenceFilter', 'unreviewed']) {
    $(id).onchange = () => {
      if (id === 'directoryFilter') fileFilter.updateFiles();
      refreshFilters();
    };
  }
  function csvCell(value) {
    let text = String(value ?? '');
    if (/^[\s]*[=+\-@]/.test(text)) text = "'" + text;
    return '"' + text.replaceAll('"', '""') + '"';
  }
  $('export').onclick = () => {
    if (!saveDraft()) return;
    const rows = entries.map((entry) => labels.get(entry.key) || entry.row);
    const csv =
      '\ufeff' +
      [
        data.fields.join(','),
        ...rows.map((row) => data.fields.map((id) => csvCell(row[id])).join(','))
      ].join('\r\n') +
      '\r\n';
    const url = URL.createObjectURL(new Blob([csv], { type: 'text/csv;charset=utf-8' }));
    const a = document.createElement('a');
    a.href = url;
    a.download = 'review.csv';
    a.click();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
    dirty = false;
    progress();
    message('已发起下载，请保管 review.csv；可重新导入继续复核。');
  };
  function parseCSV(text) {
    text = text.replace(/^\ufeff/, '');
    let rows = [],
      row = [],
      value = '',
      quoted = false;
    for (let i = 0; i < text.length; i++) {
      const c = text[i];
      if (c === '"') {
        if (quoted && text[i + 1] === '"') {
          value += '"';
          i++;
        } else quoted = !quoted;
      } else if (c === ',' && !quoted) {
        row.push(value);
        value = '';
      } else if ((c === '\n' || c === '\r') && !quoted) {
        if (c === '\r' && text[i + 1] === '\n') i++;
        row.push(value);
        if (row.some((x) => x !== '')) rows.push(row);
        row = [];
        value = '';
      } else value += c;
    }
    if (quoted) throw Error('CSV 引号未闭合');
    if (value || row.length) {
      row.push(value);
      rows.push(row);
    }
    if (!rows.length) throw Error('CSV 为空');
    const headers = rows.shift();
    if (new Set(headers).size !== headers.length) throw Error('CSV 表头重复');
    for (const id of data.fields.slice(0, 10))
      if (!headers.includes(id)) throw Error(`CSV 缺少 ${id}`);
    return rows.map((values) => {
      if (values.length !== headers.length) throw Error('CSV 列数不一致');
      const row = {};
      headers.forEach((h, i) => {
        const v = values[i];
        row[h] = /^'[\s]*[=+\-@]/.test(v) ? v.slice(1) : v;
      });
      return row;
    });
  }
  $('import').onchange = async () => {
    const file = $('import').files[0];
    if (!file) return;
    try {
      if (!saveDraft()) return;
      if (file.size > 20 * 1024 * 1024) throw Error('复核 CSV 超过 20 MiB');
      const rows = parseCSV(await file.text()),
        incoming = new Map(),
        seen = new Set();
      for (const row of rows) {
        const k = key(row);
        if (seen.has(k)) throw Error('CSV 存在重复机会');
        seen.add(k);
        const entry = byKey.get(k);
        if (!entry) throw Error('CSV 包含不属于本批次的机会');
        if (
          row.audio_sha256 !== entry.row.audio_sha256 ||
          ['at_s', 'observed_until_s'].some(
            (id) =>
              row[id] === '' ||
              !Number.isFinite(Number(row[id])) ||
              Math.abs(Number(row[id]) - entry.row[id]) > 1e-6
          )
        )
          throw Error('CSV 音频摘要或标注窗口不匹配');
        if (!row.decision) {
          if (
            [row.reviewer, row.reviewed_at, row.notes, ...extra.map((id) => row[id])].some((x) =>
              x?.trim()
            )
          )
            throw Error('CSV 有半填标签');
          continue;
        }
        validate(row);
        const old = labels.get(k);
        if (old && data.fields.some((id) => String(old[id] ?? '') !== String(row[id] ?? '')))
          throw Error('导入与本页已有标签冲突；请先导出并在重新打开的页面导入');
        incoming.set(k, row);
      }
      incoming.forEach((row, k) => labels.set(k, row));
      const wasDirty = dirty;
      formDirty = false;
      refresh();
      reconcileSelection();
      if (current) select(current);
      dirty = wasDirty;
      progress();
      message(`已导入 ${incoming.size} 条标签。`);
    } catch (e) {
      message(e.message);
    } finally {
      $('import').value = '';
    }
  };
  document.addEventListener('keydown', (event) => {
    if (/INPUT|SELECT|TEXTAREA/.test(event.target.tagName)) return;
    if (event.code === 'Space') {
      event.preventDefault();
      playback.toggle();
    } else if (event.key === 'ArrowRight') {
      event.preventDefault();
      move(1);
    } else if (event.key === 'ArrowLeft') {
      event.preventDefault();
      move(-1);
    } else if (/^[1-5]$/.test(event.key) && current) {
      $('decision').selectedIndex = Number(event.key);
      formDirty = true;
      progress();
    }
  });
  window.addEventListener('beforeunload', (event) => {
    if (dirty || formDirty) {
      event.preventDefault();
      event.returnValue = '';
    }
  });
  refresh();
  if (entries.length) select(entries[0]);
  else message('本批次没有可复核的应答机会；请在静态报告查看证据不足或处理错误。');
})();
